### DISCUSSION TITLE
Let Widgets carry an ID?

### DISCUSSION THREAD

**Discussion by laeubi** [eclipse-platform/eclipse.platform.swt#549] [Ideas]
Today I wanted to identify a widget in a set of items and wondered how this is archived and noticed there are several frameworks using different data names, e.g E4-CSS, SWT-Bot, ...

What I'm wondering is, if we should not add a generic `getId`/`setId` in the widget class so one can set an ID cross-framework (a framework can still first look for its specific id and then fallback to the generic one) to support such use-cases, the default value can simply be `null` and might be stored in an own field or reusing the `setData/getData` facility.

**Comment by akurtakov** [comment]
IMHO setData/getData is already suited for this. The issue I can think of is duplicated ids and if the implementation is on the widget level nothing will prevent having 100s of "button" ids. 
With that said I understand that it might be fine to have same id from CSS POV  (style the same way) but not from testing POV (identify exact widget). 
Clear definition of what "id" exactly means and how it would work in multiple such libraries before extra API could be done.
 What's best is that setData/getData already allows that and if/when there is agreement between multiple downstreams on the topic some "convenience API" on top of it could be added.

**Reply by laeubi** [reply]
Yes the problem is that there is no common sense of an ID, so there are multiple (maybe overlapping or even different one), so the idea would be to have a method `getId` (will mostly used by tools like e4/css or swtbot or ...) and `setId` (used by the UI devloper) and probably saving this under the key `swt.widget.id` that way it is clear (without the need of Helper function) how to define a ID for a widget in pure swt that everyone can easily understand from the code.

If/how tools then reuse this, is a another topic, a test-tool *might* first examine the "swtbot.widget.test.id" and then fallback to the widget id, while a CSS rendering might first look at the widget id and then fall back to a specific "e4.widget.css.id", but I think the main point is that a developer of a SWT application can assign a id without the need of having a special consumer in mind (or even need to set the ID twice for different purpose under different keys).

**Reply by akurtakov** [reply]
What about the case where dialog developer sets id, css engine overrides it and testing framework overrites it yet again? This doesn't sound fine although allowing only single call to setId could also be problematic.

**Reply by laeubi** [reply]
For overriding CSS+Test will use a (framework specific) property, not the generic set/getId.

**Reply by iloveeclipse** [reply]
In theory this sounds reasonable, but in practice I see the problem of responsibility for **THE** generic "ID".
If it is not set by default, it is not useful. If it is set, it adds an overhead for both computing / carrying the object ID (id's has to be unique).

As you've noticed, `getData(String)` is already doing part of that work, leaving users to implement whatever they want on top.
So not clear for me what the get/setId that are not implemented by default should do.

**Reply by laeubi** [reply]
> but in practice I see the problem of responsibility for THE generic "ID"

In fact the UI developer is responsible for define an ID (if any) if it is used in a larger scale of course there should be some kind of organization, but I don't see an issue here as we have the same concepts already on different places (package names, bundle and feature ids, products, extensionpoints, you name it..)

> If it is not set by default, it is not useful.

I don't think so, usually IDs are more used for specific things (e.g. finding a special widget, or getting the root), but if one wan't SWT can for sure generate an id (or more a path) like XML Selectors that would be for example `Shell[1]_Composite[3]_Label` ... but I don't think such automatic IDs are really useful as such "traversal ids" are already possible with all those frameworks.

> id's has to be unique

An ID *can* be unique but this is not a strict requirement (and I won't impose such a requirement here), e.g. given you have a set of labels it is often sufficient that the ID is unique in this set, but global uniqueness is not required. Also I'm not aware that any of the frameworks currently using such ID actually validates global uniqueness.

> As you've noticed, getData(String) is already doing part of that work, leaving users to implement whatever they want on top.
> So not clear for me what the get/setId that are not implemented by default should do.

The problem is that there is no common "standard" so we have a myriad of competing things that can't be used in general purpose context, I always need to be specific to a specific use-case. In contrast for example AWT has `java.awt.Component.setName(String)`, that can be used by Testing frameworks or whatever else.

Just assume I want to make E4 Applications more test-able, I could came up with a request that each SWT component created from a model object should inherit its model-id (what would be valid as E4 creates the component, so it "owns" it and can decide that the model id is a unique enough identifier for the purpose of E4 framework).

So I would open a PR for this and let is set `org.eclipse.swtbot.widget.key` ... I'm sure the first reaction would be "Don't do that this is swtbot specific and E4 must stay generic", well even if we can agree on that, what about reddeer? So probably then there would be a request to also set  `org.eclipse.reddeer.id` ... for sure it will be even more complains like "as last time said this is a bad idea an now we have the salad" ... not talking about probably other use-case.

Just now with a generic ID, one can easily agree to set the ID, as its all SWT standard, and now I can ask `reddeer`, `swt bot`, ... to please take into account the generic ID as well (as a fallback, as a primary source of truth, configurable, whatever suits that framework), and I'm not forced to set different (hopefully never changing) magic strings in my general purpose framework.

**Comment by akurtakov** [comment]
Can you try getting swtbot, reddeer, css devs to jump in here so we can hear that such an approach would be beneficial to them and they would use it? 
I ask that because the native toolkits do not provide such "id" functionality exactly for the reasons that everyone ends up overriding with their own specifics.

**Reply by merks** [reply]
If it were me, I would rather keep my IDs separate from everyone else's IDs rather than try to reuse the IDs that someone else might feel compelled to change out from under me.  E.g., we use IDs in our documentation generator and need to identify a specific widget based on our criterion.  One group might decide that a UUID would be best, while another might only care for an ID to be unique without a container, not across the entire set of widgets.  Or maybe others need it to be only unique per window.  And so on.  As far as I can tell, the only benefit of using a field would be to save space, but that seems of dubious benefit.  I expect that most cases are like my case where I want to assign the ID when I create the widget because I know that widget's purpose and need to find that purposed widget later; I imagine the CSS support is like that...

**Comment by basilevs** [comment]
Is not Java Object identity an universal identifier? With `WeakReference` there are no concerns about holding memory, it does not introduce any kind of compatibility concerns, very generic, can be used in (Weak)HashMap, immutable, automatically initialized. Perfect.

**Reply by laeubi** [reply]
> Is not Java identity an universal identifier? 

Can you explain in more details?

**Reply by basilevs** [reply]
>>Is not Java identity an universal identifier?

>Can you explain in more details?

If at some point, your code have seen a Java object of interest, it can store a reference to it:

```java
Object created = new Object();
....
Object reference = created;
```
It can then  be used to identify the object anywhere:
```java
if (unknown == reference) {
  System.out.println("Detected");
}
...

List<Object> items;
...

int index = items.indexOf(reference); // uses equals, not identity for search, but it does not matter for Widgets, as they employ default identity equality.
```

This approach suffers from memory leaks, but that can be worked around by using a `WeakReference` instead of strong references:

```java
WeakReference<Object> reference = new WeakReference<>(created);
...
int index = items.indexOf(reference.get());
```

So, as long as there is no need to persist `Widget` somehow (which is not supported, AFAIK), no other identifier is needed.

Debuggers help to identify Java identity with their own numeric ID mapping. If a text representation is needed, `System.identityHashCode()` can be used (but it is not an ID, as hash will have collisions).

**Reply by laeubi** [reply]
The point of the ID is that you define it (and reference it) without knowing the object beforehands, if I would have the object at hand I don't need an id for it :-)

**Reply by basilevs** [reply]
How do you define an object without getting a reference to it? If there was a need for a String ID, for example, at what point would you assign it? And how would you assign it without knowing a reference?
